src/dali/plugin/pair_converter/ccxt.py

```diff
@@ -0,0 +1,354 @@
+            self.__logger.debug("No path found for %s to %s. Please open an issue at %s.", from_asset, to_asset, self.ISSUES_URL)
+            return None
+
+        self.__logger.debug("Found path - %s", pricing_path)
+
+        conversion_route: List[AssetPairAndHistoricalPrice] = []
+        last_node: Optional[str] = None
+        hop_bar: Optional[HistoricalBar] = None
+
+        # Build conversion stack, we will iterate over this to find the price for each conversion
+        # Then multiply them together to get our final price.
+        for node in pricing_path:
+
+            if last_node:
+                conversion_route.append(
+                    AssetPairAndHistoricalPrice(
+                        from_asset=last_node,
+                        to_asset=node,
+                        exchange=current_markets[(last_node + node)][0],
+                        historical_data=None,
+                    )
+                )
+
+            last_node = node
+
+        for i, hop_data in enumerate(conversion_route):
+            if self._is_fiat_pair(hop_data.from_asset, hop_data.to_asset):
+                hop_bar = self._get_fiat_exchange_rate(timestamp, hop_data.from_asset, hop_data.to_asset)
+            else:
+                hop_bar = self.find_historical_bar(hop_data.from_asset, hop_data.to_asset, timestamp, hop_data.exchange)
+
+            if hop_bar is not None:
+                # Replacing an immutable attribute
+                conversion_route[i] = conversion_route[i]._replace(historical_data=hop_bar)
+            else:
+                self.__logger.debug(
+                    """No pricing data found for hop. This could be caused by airdropped
+                    coins that do not have a market yet. Market - %s%s, Timestamp - %s, Exchange - %s""",
+                    hop_data.from_asset,
+                    hop_data.to_asset,
+                    timestamp,
+                    hop_data.exchange,
+                )
+
+            if result is not None:
+                # TO BE IMPLEMENTED - override Historical Bar * to multiply two bars?
+                result = HistoricalBar(
+                    duration=max(result.duration, hop_bar.duration),  # type: ignore
+                    timestamp=timestamp,
+                    open=(result.open * hop_bar.open),  # type: ignore
+                    high=(result.high * hop_bar.high),  # type: ignore
+                    low=(result.low * hop_bar.low),  # type: ignore
+                    close=(result.close * hop_bar.close),  # type: ignore
+                    volume=(result.volume + hop_bar.volume),  # type: ignore
+                )
+            else:
+                result = hop_bar
+
+        if self.__transactions_processed % _CACHE_INTERVAL == 0:
+            self.save_historical_price_cache()
+            self.__logger.debug("Resolved %s transactions. Saving to cache.", self.__transactions_processed)
+
+        self.__transactions_processed += 1

```

Comment:
Modifying object state during computation generates a potential critical region if we ever decide to make this code multi-threaded in the future. It's not a blocker, but I was wondering if we really need this.

Reply:
I need it in order to make my run. It takes 9 hours because of the Kraken bottleneck. It saves a lot of headaches if I can save periodically.

With some speed improvements like reading in historical data from CSV, we could eliminate the bottleneck and this would mainly be unnecessary.